File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/downtime.ts
// Downtime crafting projects: create, advance days worked, and deposit the
// finished item into campaign inventory on completion. Persistent
// (`campaign_crafting`).
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject, isValidInt, SLUG_RE } from "../validation.js";
import { hasCampaign, hasCampaignCharacter } from "./campaigns.js";
import { addInventoryItem } from "./inventory.js";

interface CraftingRecord {
  campaignId: string;
  id: string;
  characterId: string;
  itemSlug: string;
  daysRequired: number;
  daysCompleted: number;
  costGp: number;
  status: string;
}

function getCraftingProject(campaignId: string, id: string): CraftingRecord | undefined {
  const row = db
    .prepare(
      "SELECT campaign_id, id, character_id, item_slug, days_required, days_completed, cost_gp, status FROM campaign_crafting WHERE campaign_id = ? AND id = ?",
    )
    .get(campaignId, id) as
    | {
        campaign_id: string;
        id: string;
        character_id: string;
        item_slug: string;
        days_required: number;
        days_completed: number;
        cost_gp: number;
        status: string;
      }
    | undefined;
  if (!row) return undefined;
  return {
    campaignId: row.campaign_id,
    id: row.id,
    characterId: row.character_id,
    itemSlug: row.item_slug,
    daysRequired: row.days_required,
    daysCompleted: row.days_completed,
    costGp: row.cost_gp,
    status: row.status,
  };
}

function saveCraftingProject(project: CraftingRecord): void {
  db.prepare(
    "INSERT INTO campaign_crafting (campaign_id, id, character_id, item_slug, days_required, days_completed, cost_gp, status) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
  ).run(
    project.campaignId,
    project.id,
    project.characterId,
    project.itemSlug,
    project.daysRequired,
    project.daysCompleted,
    project.costGp,
    project.status,
  );
}

function updateCraftingProgress(project: CraftingRecord): void {
  db.prepare("UPDATE campaign_crafting SET days_completed = ?, status = ? WHERE campaign_id = ? AND id = ?").run(
    project.daysCompleted,
    project.status,
    project.campaignId,
    project.id,
  );
}

export function handleCreateCraftingProject(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (
    !isPlainObject(body) ||
    typeof body.id !== "string" ||
    !body.id ||
    typeof body.character_id !== "string" ||
    !body.character_id ||
    typeof body.item_slug !== "string" ||
    !isValidInt(body.days_required, 1, Number.MAX_SAFE_INTEGER) ||
    !isValidInt(body.cost_gp, 0, Number.MAX_SAFE_INTEGER)
  ) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (!SLUG_RE.test(body.item_slug)) {
    sendJson(res, 400, { error: "invalid item_slug" });
    return;
  }

  if (!hasCampaignCharacter(campaignId, body.character_id)) {
    sendJson(res, 404, { error: "character not found" });
    return;
  }

  if (getCraftingProject(campaignId, body.id)) {
    sendJson(res, 409, { error: "crafting project already exists" });
    return;
  }

  const project: CraftingRecord = {
    campaignId,
    id: body.id,
    characterId: body.character_id,
    itemSlug: body.item_slug,
    daysRequired: body.days_required,
    daysCompleted: 0,
    costGp: body.cost_gp,
    status: "active",
  };
  saveCraftingProject(project);

  sendJson(res, 201, {
    id: project.id,
    character_id: project.characterId,
    item_slug: project.itemSlug,
    days_required: project.daysRequired,
    days_completed: project.daysCompleted,
    status: project.status,
  });
}

export function handleAdvanceCrafting(
  res: ServerResponse,
  campaignId: string,
  projectId: string,
  body: unknown,
): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const project = getCraftingProject(campaignId, projectId);
  if (!project) {
    sendJson(res, 404, { error: "crafting project not found" });
    return;
  }

  if (!isPlainObject(body) || !isValidInt(body.days, 1, Number.MAX_SAFE_INTEGER)) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  if (project.status === "complete") {
    sendJson(res, 409, { error: "crafting project already complete" });
    return;
  }

  project.daysCompleted = Math.min(project.daysCompleted + body.days, project.daysRequired);
  const justCompleted = project.daysCompleted >= project.daysRequired;
  project.status = justCompleted ? "complete" : "active";
  updateCraftingProgress(project);

  if (justCompleted) {
    addInventoryItem(campaignId, project.itemSlug, "party", 1, false);
  }

  sendJson(res, 200, {
    id: project.id,
    days_completed: project.daysCompleted,
    status: project.status,
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T060905Z_claude_claude-sonnet-5_python-django/dndsite/views/downtime.py
"""Downtime crafting projects, keyed off a campaign_id."""

import json

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from dndsite import db
from dndsite.views._util import json_body


@csrf_exempt
def campaign_crafting_collection(request, campaign_id):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)

    if db.get_campaign(campaign_id) is None:
        return JsonResponse({"error": "campaign not found"}, status=404)

    try:
        body = json_body(request)
        project_id = body["id"]
        character_id = body["character_id"]
        item_slug = body["item_slug"]
        days_required = body["days_required"]
        cost_gp = body["cost_gp"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not isinstance(project_id, str) or not project_id:
        return JsonResponse({"error": "invalid id"}, status=400)
    if not isinstance(character_id, str) or not character_id:
        return JsonResponse({"error": "invalid character_id"}, status=400)
    if not isinstance(item_slug, str) or not item_slug:
        return JsonResponse({"error": "invalid item_slug"}, status=400)
    if not isinstance(days_required, int) or isinstance(days_required, bool) or days_required <= 0:
        return JsonResponse({"error": "invalid days_required"}, status=400)
    if not isinstance(cost_gp, int) or isinstance(cost_gp, bool) or cost_gp < 0:
        return JsonResponse({"error": "invalid cost_gp"}, status=400)

    if db.get_campaign_character(campaign_id, character_id) is None:
        return JsonResponse({"error": "character not found"}, status=404)

    if db.get_crafting_project(campaign_id, project_id) is not None:
        return JsonResponse({"error": "crafting project already exists"}, status=409)

    project = {
        "id": project_id,
        "character_id": character_id,
        "item_slug": item_slug,
        "days_required": days_required,
        "days_completed": 0,
        "cost_gp": cost_gp,
        "status": "active",
    }
    db.create_crafting_project(campaign_id, project)

    return JsonResponse(
        {
            "id": project["id"],
            "character_id": project["character_id"],
            "item_slug": project["item_slug"],
            "days_required": project["days_required"],
            "days_completed": project["days_completed"],
            "status": project["status"],
        },
        status=201,
    )


@csrf_exempt
def campaign_crafting_advance(request, campaign_id, project_id):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)

    if db.get_campaign(campaign_id) is None:
        return JsonResponse({"error": "campaign not found"}, status=404)

    project = db.get_crafting_project(campaign_id, project_id)
    if project is None:
        return JsonResponse({"error": "crafting project not found"}, status=404)

    try:
        body = json_body(request)
        days = body["days"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    if not isinstance(days, int) or isinstance(days, bool) or days <= 0:
        return JsonResponse({"error": "invalid days"}, status=400)

    if project["status"] != "active":
        return JsonResponse({"error": "crafting project not active"}, status=409)

    project["days_completed"] = min(project["days_completed"] + days, project["days_required"])
    if project["days_completed"] >= project["days_required"]:
        project["status"] = "complete"

    db.save_crafting_project(campaign_id, project)

    if project["status"] == "complete":
        db.add_inventory_item(campaign_id, project["item_slug"], "party", 1)

    return JsonResponse(
        {
            "id": project["id"],
            "days_completed": project["days_completed"],
            "status": project["status"],
        }
    )
